openssl/src/ssl/callbacks.rs

```diff
@@ -416,3 +424,82 @@ where
+                                        _: *mut c_void)
+                                        -> c_int
+    where F: FnMut(&mut SslRef, ExtensionContext, Option<(usize, &mut X509Ref)>) -> Result<Option<Cow<'static, [u8]>>, SslAlert> + 'static
+{
+    unsafe {
+        let ssl_ctx = ffi::SSL_get_SSL_CTX(ssl as *const _);
+        let callback_data = ffi::SSL_CTX_get_ex_data(ssl_ctx, get_callback_idx::<CustomExtAddState<F>>());
+        let callback_data = &mut *(callback_data as *mut CustomExtAddState<F>);
+        let ssl = SslRef::from_ptr_mut(ssl);
+        let ectx = ExtensionContext::from_bits(context).unwrap();
+        let cert = if ectx == ExtensionContext::TLS1_3_CERTIFICATE { Some((chainidx, X509Ref::from_ptr_mut(x))) } else { None };
+        match (callback_data.0)(ssl, ectx, cert) {
+            Ok(None) => 0,
+            Ok(Some(buf)) => {
+                *outlen = buf.len() as size_t;
+                *out = buf.as_ptr();
+                callback_data.1 = Some(buf);
+                1
+            }
+            Err(alert) => {
+                *al = alert.0;
+                -1
+            }
+        }
+    }
+}
+
+#[cfg(all(feature = "v111", ossl111))]
+pub extern "C" fn raw_custom_ext_free<F>(ssl: *mut ffi::SSL, _: c_uint,
+                                         _: c_uint,
+                                         _: *mut *const c_uchar,
+                                         _: *mut c_void)
+    where F: 'static
+{
+    unsafe {
+        let ssl_ctx = ffi::SSL_get_SSL_CTX(ssl as *const _);
+        let callback_data = ffi::SSL_CTX_get_ex_data(ssl_ctx, get_callback_idx::<(F, Option<Cow<'static, [u8]>>)>());
+        let data = &mut (*(callback_data as *mut (F, Option<Cow<'static, [u8]>>))).1;
+        data.take();
+    }
+}
+
+#[cfg(all(feature = "v111", ossl111))]
+pub extern "C" fn raw_custom_ext_parse<F>(ssl: *mut ffi::SSL, _: c_uint,
+                                          context: c_uint,
+                                          input: *const c_uchar,
+                                          inlen: size_t, x: *mut ffi::X509,
+                                          chainidx: size_t, al: *mut c_int,
+                                          _: *mut c_void)
+                                        -> c_int
+    where F: FnMut(&mut SslRef, ExtensionContext, &[u8], Option<(usize, &mut X509Ref)>) -> Result<(), SslAlert> + 'static
+{
+    unsafe {
+        let ssl_ctx = ffi::SSL_get_SSL_CTX(ssl as *const _);
+        let callback = ffi::SSL_CTX_get_ex_data(ssl_ctx, get_callback_idx::<F>());
+        let ssl = SslRef::from_ptr_mut(ssl);
+        let callback = &mut *(callback as *mut F);
+        let ectx = ExtensionContext::from_bits(context).unwrap();
+        let slice = slice::from_raw_parts(input as *const u8, inlen as usize);
+        let cert = if ectx == ExtensionContext::TLS1_3_CERTIFICATE { Some((chainidx, X509Ref::from_ptr_mut(x))) } else { None };
```

Comment:
shouldn't this be `if ectx.contains(TLS1_3_CERTIFIACTE)` since ExtensionContext is a bitfield?

Reply:
Because this indicates the message type being constructed/parsed, I don't expect it will ever contain multiple values. That said, if I'm right it doesn't matter either way, and if I'm wrong then `contains` should be used, so I'll switch to hedge our bets.